core/src/main/scala/org/apache/spark/storage/disk/DiskBlockObjectWriter.scala

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage.disk
+
+import java.io.File
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.serializer.{SerializationStream, SerializerInstance}
+import org.apache.spark.storage.BlockId
+import org.apache.spark.util.Utils
+
+/**
+ * A class for writing JVM objects directly to a file on disk. This class allows data to be appended
+ * to an existing block and can guarantee atomicity in the case of faults as it allows the caller to
+ * revert partial writes.
+ *
+ * This class does not support concurrent writes. Also, once the writer has been opened it cannot be
+ * reopened again.
+ */
+private[spark] class DiskBlockObjectWriter(
+    private[this] var diskBlockWriter: DiskBlockWriter,
+    serializerInstance: SerializerInstance,
+    val blockId: BlockId = null)
+  extends Logging {
+
+  private var objOut: SerializationStream = null
+  private var initialized = false
+  private var hasBeenClosed = false
+  private var commitAndCloseHasBeenCalled = false
+
+  def file: File = diskBlockWriter.file
+
+  def open(): DiskBlockObjectWriter = {
+    if (hasBeenClosed) {
+      throw new IllegalStateException("Writer already closed. Cannot be reopened.")
+    }
+    diskBlockWriter.open()
+    objOut = serializerInstance.serializeStream(diskBlockWriter)
+    initialized = true
+    this
+  }
+
+  def close() {
+    Utils.tryWithSafeFinally {
+      if (initialized) {
+        if (diskBlockWriter.syncWrites) {
+          objOut.flush()
+        }
+      }
+    } {
+      if (initialized) {
+        objOut.close()
+      }
+      diskBlockWriter = null
+      objOut = null
+      initialized = false
+      hasBeenClosed = true
+    }
+  }
+
+  /**
+   * Flush the partial writes and commit them as a single atomic block.
+   */
+  def commitAndClose(): Unit = {
+    if (initialized) {
+      objOut.flush()
```

Comment:
Update: actually, the old code called `objOut.close()` as part of `close()`, so we still need to figure out how to make sure that happens here :(